File: Data Structures & Algorithms/subsets/submission-1.java
class Solution {

    public void solve(int index, List<Integer> tmp, List<List<Integer>> ans, int[] nums){
        if(index == nums.length){
            ans.add(new ArrayList<>(tmp));
            return;
        }

        tmp.add(nums[index]);
        solve(index+1, tmp, ans, nums);
        tmp.remove(tmp.size()-1);
        solve(index+1, tmp, ans, nums);
    }

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> tmp = new ArrayList<>();

        solve(0, tmp, ans, nums);
        return ans;
    }
}

File: Data Structures & Algorithms/house-robber/submission-1.cpp
class Solution {
public:

    int solve(int i, vector<int>& cost, vector<int>& dp){
        if(i < 0)
            return 0;

        if(dp[i] != INT_MIN)
            return dp[i];

        int a = cost[i] + solve(i-2, cost, dp);
        int b = solve(i-1, cost, dp);

        return dp[i] = max(a,b);
    }

    int rob(vector<int>& nums) {
        if(nums.size() == 1)
            return nums[0];

        vector<int> dp(nums.size(), INT_MIN);

        return solve(nums.size()-1, nums, dp);
    }
};
